File: packages/ui/db/entity/Photoable.ts
import { Area, Boulder, Photo, Route } from "@/db";
import { cascadeOneToMany } from "@/db/cascadeOptions";
import BaseColumnSchemaPart from "@/db/entity/BaseColumnSchemaPart";
import { Crag } from "@/db/entity/Crag";
import { IPhotoable } from "models";
import { EntitySchema } from "typeorm";

export type Photoable = IPhotoable & {
  photos: Photo[];
  crag?: Crag;
  area?: Area;
  boulder?: Boulder;
  route?: Route;
};

const PhotoableSchema = new EntitySchema<Photoable>({
  name: "photoable",
  columns: {
    ...BaseColumnSchemaPart,
    descriptor: {
      type: String,
    },
  },
  relations: {
    photos: {
      type: "one-to-many",
      target: "photo",
      inverseSide: "photoable",
      ...cascadeOneToMany,
    },
    crag: {
      type: "one-to-one",
      target: "crag",
      inverseSide: "photoable",
    },
    area: {
      type: "one-to-one",
      target: "area",
      inverseSide: "photoable",
    },
    boulder: {
      type: "one-to-one",
      target: "boulder",
      inverseSide: "photoable",
    },
    route: {
      type: "one-to-one",
      target: "route",
      inverseSide: "photoable",
    },
  },
});

export default PhotoableSchema;

File: go-server/models/photoable.go
package models

type Photoable struct {
	ID         *int     `json:"id,omitzero"`
	Crag       *Crag    `json:"crag,omitzero"`
	Area       *Area    `json:"area,omitzero"`
	Boulder    *Boulder `json:"boulder,omitzero"`
	Route      *Route   `json:"route,omitzero"`
	Descriptor string   `json:"descriptor"`
}

func (p Photoable) IsZero() bool {
	return p.ID == nil && p.Crag == nil && p.Area == nil && p.Boulder == nil && p.Route == nil && p.Descriptor == ""
}
